docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+Both `Count` and `IsHardwareAccelerated` are turned into constants by the Just-In-Time compiler (i.e. no method call is required to retrieve the information). In the case of pre-compiled code (NativeAOT), this is not true for the `IsHardwareAccelerated` property, as the required information is not available at compile time.
+
+**Note:** When `Vector256` is accelerated then `Vector128` and `Vector64` are also accelerated.
+
+That is why the code is very often structured like this:
+
+```cs
+void CodeStructure(ReadOnlySpan<byte> buffer)
+{
+    if (Vector256.IsHardwareAccelerated && buffer.Length >= Vector256<byte>.Count)
+    {
+        // Vector256 code path
+    }
+    else if (Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<byte>.Count)
+    {
+        // Vector128 code path
+    }
+    else
+    {
+        // non-vectorized && small inputs code path
+    }
+}
+```
+
+To reduce the number of comparisons for small inputs, we can re-arrange it in the following way:
+
+```cs
+void OptimalCodeStructure(ReadOnlySpan<byte> buffer)
+{
+    if (!Vector128.IsHardwareAccelerated || buffer.Length < Vector128<byte>.Count)
+    { 
+        // scalar code path
+    } 
+    else if (!Vector256.IsHardwareAccelerated || buffer.Length < Vector256<byte>.Count)
+    { 
+        // Vector128 code path
+    } 
+    else
+    { 
+        // Vector256 code path
+    }
+}
+```
+
+**Both vector types provide almost identical features**, but arm64 hardware does not support `Vector256` yet, so for the sake of simplicity we will be using `Vector128` in all examples and assuming **little endian** architecture. Which means that all examples used in this document assume that they are being executed as part of the following `if` block:
+
+```cs
+else if (Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<byte>.Count)
+{
+    // Vector128 code path
+}
+```
+
+### Testing
+
+Such a code structure requires us to **test all possible code paths**:
+
+* `Vector256` is accelerated:
+  * The input is large enough to benefit from vectorization with `Vector256`.
+  * The input is not large enough to benefit from vectorization with `Vector256`, but it can benefit from vectorization with `Vector128`.
+  * The input is too small to benefit from any kind of vectorization.
+* `Vector128` is accelerated
+  * The input is large enough to benefit from vectorization with `Vector128`.
+  * The input is too small to benefit from any kind of vectorization.
+* Neither `Vector128` or  `Vector256` are accelerated.
+
+It's possible to implement tests that cover some of the scenarios based on the size, but it's impossible to toggle hardware acceleration at the unit test level. It can be controlled with environment variables before .NET process is started:
```

Comment:
It would be relatively simple for us to expose a `DOTNET_EnableVector128=0` (and same for `Vector<T>`, `Vector64/256/512<T>`) knob. Could you open an issue suggesting it?

We'd need to callout that its .NET 8 only, but this would help with testing, particularly across platforms.